max nb columns for print : more cleaning

guilpier-code opened the pull request:
The current PR follows #1172.
This is a cleaning work.

**All PR about max nb columns for print** : 
There are currently 3 PR about  **max nb columns for print**
- #1159 : some preliminary cleaning
- #1172 : adresses the real problem : computing this max number of columns accurately
- The current PR

**What this PR is about** :
The first target was to replace an **std::vector** with a **std::map** in class **AllVariablesPrintInfo** and hope it would simplify the code. As a matter of fact, it does.
What was done actually : 
-  replacing an **std::vector** with a **std::map** in class **AllVariablesPrintInfo**, it simplifies methods **setPrintStatus(...)**, **setMaxColumns(...)** and **isPrinted(...)**
- In this **std::map** (that associates **name of vars** to instance of **VariablesPrintInfo**), avoid inserting **VariablesPrintInfo** as pointers. This allows to make the destructor of  **AllVariablesPrintInfo** empty.
- Introducing some more methods to class **AllVariablesPrintInfo** to simplify the code elsewhere, see : 
     - **src/libs/antares/study/parameters.cpp**
     - 	**src/ui/simulator/toolbox/components/datagrid/renderer/select-variables.cpp**
     - **src/ui/simulator/windows/options/select-output/select-output.cpp**
- some more renaming and comments left in the code

guilpier-code committed:
max nb columns for print : code improvement

Improvements :
- At reading the output variable print status (parameters.cpp)
- Small changes inside classes of variable-print-info.h/.cpp
- Removing function SurveyResults::initializeMaxVariables(...), which has shrunk until it does one function call.

guilpier-code committed:
max nb columns for print : in class AllVariablesPrintInfo, changing the std::vector into a std::map

guilpier-code committed:
max nb columns for print : renaming in output variable BC MARG COST

flomnes commented on src/libs/antares/study/variable-print-info.cpp:
```diff
@@ -83,120 +82,110 @@ void variablePrintInfoCollector::add(const AnyString& name,
+void AllVariablesPrintInfo::reverseAll()
+{
+    for (auto pair : allVarsPrintInfo)
+        pair.second.reverse();
```
```suggestion
    for (auto& pair : allVarsPrintInfo)
        pair.second.reverse();
```
Otherwise you'd change a copy, not the elements themselves. See [this example](https://gist.github.com/flomnes/008920c58ba607c91d201b42a543507c).

guilpier-code replied:
Thanks, this would have caused bugs surely !
Corrected now.

guilpier-code committed:
max nb columns for print : fix a bug

guilpier-code committed:
max nb columns for print : fix other bugs

flomnes commented on src/libs/antares/study/variable-print-info.cpp:
```diff
@@ -220,15 +219,15 @@ void AllVariablesPrintInfo::computeMaxColumnsCountInReports()
         {
             uint currentColumnsCount = 0;
-            for (auto it = allVarsPrintInfo.begin(); it != allVarsPrintInfo.end(); it++)
+            for (auto& pair : allVarsPrintInfo)
```
C++17 offers the more concise syntax `for(auto& [key, value] : myMap)`. Make sure to use it with expressive names for `key` and `value`.

guilpier-code replied:
More readable indeed !
Suggestion applied.

flomnes commented on src/libs/antares/study/variable-print-info.cpp:
```diff
@@ -239,18 +238,18 @@ void AllVariablesPrintInfo::computeMaxColumnsCountInReports()
-        if (v->isPrinted() && v->getDataLevel() == Category::link)
+        if (pair.second.isPrinted() && pair.second.getDataLevel() == Category::link)
             numberSelectedLinkVariables++;
     }
```
https://en.cppreference.com/w/cpp/algorithm/count

guilpier-code replied:
Here have to choose between use of std algorithm and readability : 
```C++
numberSelectedAreaVariables = std::count_if(
        allVarsPrintInfo.begin(),
        allVarsPrintInfo.end(),
        [](auto& p) {return p.second.isPrinted() &&
                            p.second.getDataLevel() == Category::area; });
```
or 
```C++
    for (auto& [name, variable] : allVarsPrintInfo)
    {
       if (variable.isPrinted() && variable.getDataLevel() == Category::area)
           numberSelectedAreaVariables++;
    }
```

flomnes commented on src/libs/antares/study/variable-print-info.cpp:
```diff
@@ -83,120 +82,120 @@ void variablePrintInfoCollector::add(const AnyString& name,
-
     // Initializing output variables status
-    if (!userSelection)
+    if (not isThematicTrimmingEnabled)
```
```suggestion
    if (!isThematicTrimmingEnabled)
```

flomnes commented on src/libs/antares/study/variable-print-info.cpp:
```diff
@@ -83,120 +82,120 @@ void variablePrintInfoCollector::add(const AnyString& name,
 // ============================================================
 
-AllVariablesPrintInfo::~AllVariablesPrintInfo()
+static std::string to_uppercase(std::string str)
```
```suggestion
static std::string to_uppercase(const std::string& str)
```

guilpier-code replied:
see answer below

flomnes commented on src/libs/antares/study/variable-print-info.cpp:
```diff
@@ -83,120 +82,120 @@ void variablePrintInfoCollector::add(const AnyString& name,
 }
 
-void AllVariablesPrintInfo::add(VariablePrintInfo* v)
+void AllVariablesPrintInfo::add(std::string name, VariablePrintInfo v)
```
```suggestion
void AllVariablesPrintInfo::add(const std::string& name, VariablePrintInfo v)
```

guilpier-code replied:
Giving a **const string&** to **std::transform** does not compile. But **string&** is ok

flomnes commented on src/libs/antares/study/variable-print-info.cpp:
```diff
@@ -83,120 +82,120 @@ void variablePrintInfoCollector::add(const AnyString& name,
+    if (not exists(upperCaseName))
+    {
+        index_to_name[(unsigned int)allVarsPrintInfo.size()] = upperCaseName;
+        allVarsPrintInfo.insert(std::pair<std::string, VariablePrintInfo>(upperCaseName, v));
```
```suggestion
        allVarsPrintInfo[upperCaseName] = v;
```

guilpier-code replied:
About that, I first wrote the line : 
```C++
allVarsPrintInfo[upperCaseName] = v;
```
But it does not compile if class **VariablePrintInfo** (type of the values in the **std::map**) does not have an explicitly defined **default constructor**.
Indeed, as you may know, the **std::map [] operator** needs to add a default value in the map in case the key does have already a associated value.
As I didn't want this **default constructor** to be defined (to make code shorter), I opted for the **insert** method.

flomnes reviewed (commented):
You need strategy to keep `namesPrinted` & `namesNotPrinted` up-to-date when enabling/disabling an output variable.

Either call it by after every for-loop on adding/removing variables. I think it might be out-of-sync in some cases, for example parameters.cpp:1883.

Or (this looks better) you could use a bool `sync` flag
- At construction, `sync=true` (everything is empty, hence synchronized)
- At every write, `sync=false`
- At every read (including e.g `namesPrinted.size()`),
  - if `sync==false`, call `sortVariablesByPrintStatus` then set `sync=true`
  - if `sync==true`, don't do anything
 
You'd need to encapsulate `namesPrinted.size()` into a dedicated method.

guilpier-code committed:
max nb columns for print : refresh names lists each time a print status is changed

guilpier-code commented:
> You need strategy to keep `namesPrinted` & `namesNotPrinted` up-to-date when enabling/disabling an output variable.

This has been taken care of. Tell me what you think.

guilpier-code committed:
max nb columns for print : removing useless isEmpty()

Indeed, allVarsPrintInfo can never be empty, has it is build from code built at compile time

guilpier-code committed:
max nb columns for print : try fix a compilation failure on Linux

guilpier-code committed:
max nb columns for print : change for more readable manner to traverse std::map

guilpier-code committed:
max nb columns for print : corrections after review

guilpier-code commented on src/libs/antares/study/variable-print-info.h:
```diff
@@ -104,18 +95,21 @@ class AllVariablesPrintInfo
-    VariablePrintInfo* operator[](uint i) const;
+    VariablePrintInfo& operator[](uint i);
     size_t size() const;
-    bool isEmpty() const;
```
We don't need **isEmpty()**, since the output variables are built at compile time and their number is fixed

guilpier-code committed:
max nb columns for print : more corrections after review

flomnes reviewed (commented):
It seems that this PR kind of misses the goal of simplifying the management of printed/hidden variables.

We introduce 2 vectors (`namesOfEnabledVariables` and `namesOfDisabledVariables`) and a map `index_to_name` that need to be kept synchronized when adding/removing/reverting a variable.

Maybe we should do it differently ? Let's discuss this option

guilpier-code commented:
> It seems that this PR kind of misses the goal of simplifying the management of printed/hidden variables.
> 
> We introduce 2 vectors (`namesOfEnabledVariables` and `namesOfDisabledVariables`) and a map `index_to_name` that need to be kept synchronized when adding/removing/reverting a variable.
> 
> Maybe we should do it differently ? Let's discuss this option

I am close to agree.
That said we have small benefits :

guilpier-code committed:
max nb columns for print : amend previous commit (causes compilation error on Linux)

guilpier-code commented:
> It seems that this PR kind of misses the goal of simplifying the management of printed/hidden variables.
> 
> We introduce 2 vectors (`namesOfEnabledVariables` and `namesOfDisabledVariables`) and a map `index_to_name` that need to be kept synchronized when adding/removing/reverting a variable.
> 
> Maybe we should do it differently ? Let's discuss this option

I am close to agree.
While working on it, I was having the same concern because I saw **variable-print-info.h/.cpp** grow and get more complex.
That said, we have small benefits : 
- on this PR, we reduce globally (a tiny bit) the number of lines.
- yes, the code of **variable-print-info.h/.cpp** has grown because they offer more services to clients of this code, and makes the callers simpler.
- In classes of **variable-print-info.h/.cpp,** old methods (existing before this PR) are globally tinier and more understandable
    - **setPrintStatus**, setMaxColumns**, **clear** and **isPrinted** has shrunk
    - **AllVariablesPrintInfo** has now a default destructor

guilpier-code committed:
max nb columns for print : remove lists that have to be updated in real-time

flomnes commented on src/libs/antares/study/parameters.cpp:
```diff
@@ -1876,35 +1873,23 @@ void Parameters::saveToINI(IniFile& ini) const
-        assert(!variablesPrintInfo.isEmpty());
         uint nb_tot_vars = (uint)variablesPrintInfo.size();
-        uint selected_vars = 0;
+        uint nb_selected_vars = (uint)variablesPrintInfo.namesOfEnabledVariables().size();
```
You create a `vector` here, but use only its size. I suggest creating a dedicated method with a simple implementation (using e.g `std::count` or `std::count_if`) instead.

flomnes reviewed (commented):
That's much better. One last thing and we're good

flomnes committed:
Use enum value in template param (#1193)

guilpier-code committed:
max nb columns for print : adding member function numberOfEnabledVariables() to class AllVariablesPrintInfo
